Implement (simplified) Deployment controller and its model

Catoverflow opened the pull request:
- Add `VDeployment` controller model, implementation and documentation
- Add e2e test in CI and local test script for `VDeployment` controller
- As required for `VDeployment` controller, add
  - setter for `VReplicaSet` and its children fields
  - `Clone` trait for `VReplicaSet`
  - `well_formed` for `ObjectMeta` and `VReplicaSet`
  - `eq` for `PodTemplateSpec`
  - `Default` trait for `VReplicaSet`
  - picked `push_filter_and_filter_push` out of ad-hoc implementation from `VReplicaSet` to `vstd`
- Updated test scripts of `local-test.sh` and `deploy.sh` to separate logic for build and deploy (may be a bit out of scope)
 For e2e tests, I prefer `k8s.gcr.io/pause:3.9` over `php-redis`.
- Merge test scripts of admission controllers into `deploy.sh`

Catoverflow commented:
The current state machine implementation is hard for reasoning over termination. Should switch to another design which keeps track of old_vrs_list and new_vrs in state following the vanilla implementation.

codyjrivera commented on src/kubernetes_api_objects/exec/object_meta.rs:
```diff
@@ -186,6 +186,16 @@ impl ObjectMeta {
+    pub fn well_formed(&self) -> (b: bool)
+        ensures b == self@.well_formed(),
+    {
+        self.inner.name.is_some()
```
Super-ignorable nit: I'd personally indent this as?
```
self.inner.name.is_some()
&& self.inner.namespace.is_some()
&& self.inner.resource_version.is_some()
&& self.inner.uid.is_some()
```

codyjrivera commented on src/v2/controllers/vdeployment_controller/exec/reconciler.rs:
```diff
@@ -0,0 +1,505 @@
+        assert(forall |i: int| 0 <= i < old_vrs_list.len() ==> old_vrs_list[i]@.well_formed());
+    }
+
+    while idx < vrs_list.len()
```
Good job figuring out how to put both new and old lists into a single loop invariant!

codyjrivera commented on src/v2/controllers/vdeployment_controller/exec/reconciler.rs:
```diff
@@ -0,0 +1,505 @@
+    let template = template_with_hash(vd);
+    spec.set_template(template);
+    vrs.set_spec(spec);
+    proof {
```
Is this `proof` block really needed anymore now we fixed the dash/underscore discrepancy?

codyjrivera commented on src/v2/controllers/vdeployment_controller/exec/reconciler.rs:
```diff
@@ -0,0 +1,505 @@
+    let mut pod_template_spec = PodTemplateSpec::default();
+    pod_template_spec.set_metadata(template_meta);
+    pod_template_spec.set_spec(vd.spec().template().unwrap().spec().unwrap());
+    proof {
```
Similar to above

codyjrivera commented on src/v2/controllers/vdeployment_controller/trusted/spec_types.rs:
```diff
@@ -127,9 +124,11 @@ impl ResourceView for VDeploymentView {
         &&& self.spec.template.get_Some_0().spec.is_Some()
         // selector matches template's metadata's labels
-        &&& self.spec.selector.matches(self.spec.template.get_Some_0().metadata.get_Some_0().labels.unwrap_or(Map::empty()))
+        &&& self.spec.selector.matches(self.spec.template.get_Some_0().metadata.get_Some_0().labels.get_Some_0())
```
I'm confused about this line. Please explain.

Catoverflow replied:
Should be accidentally introduced when I tried to have corresponding part in exec and failed. Reverted

codyjrivera reviewed (commented):
A few of my comments might be a bit stale given our meeting, fyi.

Catoverflow commented on discussion/kubernetes-model/deployment_controller.md:
```diff
@@ -0,0 +1,56 @@
+
+> In k8s implementation it's [possible](https://github.com/kubernetes/kubernetes/blob/cdc807a9e849b651fb48c962cc18e25d39ec5edf/pkg/controller/deployment/util/deployment_util.go#L633-L634) to have multiple new replica sets in rare cases
+
+**: The scale up/down process should satisfy `maxSurge` and `maxUnavailable` properties of DC, which means DC may not scale the controlled new & old RS to desired replicas in one step. Instead DC will scale up & down gradually in turn. Currently we do not model this feature.
```
Yes, let me update this ambiguous description here.

Catoverflow commented on e2e/src/v2_vreplicaset_e2e.rs:
```diff
@@ -16,6 +16,11 @@ use kube::{
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 use serde_json::json;
+use deps_hack::anyhow::Result;
```
No, fixed

Catoverflow commented:
@codyjrivera Thanks for your comments! I have updated some of them, while as I'm going to rewrite the VD some implementation-specific problems will be rechecked after the rewrite is complete.

Catoverflow commented on src/v2/controllers/vdeployment_controller/model/reconciler.rs:
```diff
@@ -0,0 +1,246 @@
+            }
+        },
+        VDeploymentReconcileStepView::RollReplicas => {
+            // TODO: support different policy (order of scaling of new and old vrs)
```
I think the code makes decisions on which to update first is in this state, and it also determines how many to update (for `maxSurge` and `maxUnavailable`)

marshtompsxd commented on deploy.sh:
```diff
@@ -15,8 +15,25 @@ app=$(echo "$1" | tr '_' '-') # should be the controller's name (with words sepa
+fi
+
+# use imperative management for CRDs since metadata for PodTemplateSpec is too long.
+# why don't we directly create CRD using kube API?
```
@Catoverflow we don't want to assume what kube client lib people are using, so we use kubectl here.
btw if you want to leave a comment for future improvement, please start with `TODO: `

Catoverflow replied:
After the discussion I think it's fine, let me remove it

marshtompsxd commented on discussion/kubernetes-model/deployment_controller.md:
@Catoverflow Thank you for this document. I suggest you move it to the code file (as comments). A problem of this type of doc is that they easily get out-of-synced when the code evolves. Putting the doc with the code does not solve the problem, but makes it better. It's also easier for anyone who reads the doc to relate to the code.

Catoverflow replied:
I moved it to `src/v2/controllers/vdeployment_controller/trusted/step.rs`

marshtompsxd commented on e2e/src/v2_vreplicaset_e2e.rs:
```diff
@@ -16,6 +16,10 @@ use kube::{
+use deps_hack::serde_yaml;
+use deps_hack::tokio;
+use deps_hack::tracing::{error, info};
+use deps_hack::tracing_subscriber;
```
Do we need these new imports?

Catoverflow replied:
Removed

marshtompsxd commented on src/kubernetes_api_objects/exec/label_selector.rs:
```diff
@@ -69,7 +70,9 @@ impl LabelSelector {
-        ensures res == self@.matches(labels@)
+        ensures
+            res == self@.matches(labels@),
+            res ==> self@.match_labels.is_None() || labels@.len() >= self@.match_labels.get_Some_0().len(),
```
Why do we need this second postcondition here? Isn't it already implied by the first postcondition?

Catoverflow replied:
It's required by `VReplicaSetView.state_validation` but it's automatically proved. Let me remove it.

Catoverflow commented on e2e/src/v2_vdeployment_e2e.rs:
```diff
@@ -0,0 +1,496 @@
+
+    info!("E2e test passed.");
+    Ok(())
+}
```
That is fine I guess? We separate tests in a "code block"

Catoverflow commented on local-test.sh:
```diff
@@ -3,18 +3,71 @@
 ## Test the controller locally in a kind cluster.
 ##
 ## Requires kind to be installed and the prerequisites of deploy.sh.
+## Usage: ./local-test.sh <controller_name> [--no-build]
```
@codyjrivera I think the problem is `vargo` does not support build cache. So I can only reuse the binary as a whole

Catoverflow commented on src/kubernetes_api_objects/exec/label_selector.rs:
```diff
@@ -69,7 +70,9 @@ impl LabelSelector {
     pub fn matches(&self, labels: StringMap) -> (res: bool)
-        ensures res == self@.matches(labels@)
+        ensures
+            res == self@.matches(labels@),
```
@codyjrivera I think `labels@.len() >= self@.match_labels.get_Some_0().len()` is provable, and that's required by the `VReplicaSetView.state_validation()`, where `spec.selector` and its labels must be non-empty, and using this post-condition we know `template.metadata.labels` must be non-empty. But you are right, may be this should be written as a lemma?

Catoverflow replied:
I tried and it proved without the lemma, should be fine. Let me remove it.

Catoverflow replied:
It's indeed a legacy to be fixed. I fix it temporarily by changing the `vrs.state_validation`

Catoverflow commented:
@marshtompsxd I want to unify the test pipeline, but currently `v2-vreplicaset-admission` has its own scripts in `src/v2/controllers/vreplicaset_controller/admission_control`. I think it's better to move them to `deploy` and `e2e` so we don't need special treatment in ci and scripts. That's beyond the scope of this PR so I just added simple workaround in `deploy.sh` and changed the name in its scripts to use the same image name as other controllers

marshtompsxd commented:
> @marshtompsxd I want to unify the test pipeline, but currently `v2-vreplicaset-admission` has its own scripts in `src/v2/controllers/vreplicaset_controller/admission_control`. I think it's better to move them to `deploy` and `e2e` so we don't need special treatment in ci and scripts. That's beyond the scope of this PR so I just added simple workaround in `deploy.sh` and changed the name in its scripts to use the same image name as other controllers

I totally agree

marshtompsxd commented on src/vstd_ext/seq_lib.rs:
```diff
@@ -271,6 +271,15 @@ pub proof fn push_to_set_eq_to_set_insert<A>(s: Seq<A>, e: A)
+        !pred(e) ==> s.push(e).filter(pred) == s.filter(pred),
+{
+    reveal(Seq::filter);
+    assert(s.push(e).drop_last() == s);
```
Just curious: is the last assertion here necessary?

marshtompsxd commented on src/v2/controllers/vreplicaset_controller/trusted/spec_types.rs:
```diff
@@ -124,7 +138,8 @@ impl ResourceView for VReplicaSetView {
         &&& self.spec.template.get_Some_0().spec.is_Some()
         // selector matches template's metadata's labels
-        &&& self.spec.selector.matches(self.spec.template.get_Some_0().metadata.get_Some_0().labels.unwrap_or(Map::empty()))
+        &&& self.spec.template.get_Some_0().metadata.get_Some_0().labels.is_Some()
```
Just want to double check with you: the labels must be `Some`, is that correct?

marshtompsxd commented on src/v2/controllers/vreplicaset_controller/trusted/exec_types.rs:
```diff
@@ -131,6 +174,19 @@ pub struct VReplicaSetSpec {
+        ensures vreplicaset_spec@ == spec_types::VReplicaSetSpecView::default(),
+    {
+        VReplicaSetSpec {
+            inner: deps_hack::VReplicaSetSpec {
```
Why not directly call `inner::default()`? VReplicaSetSpec has Default trait now https://github.com/anvil-verifier/anvil/pull/611/files#diff-4b77f89f66f064750dbbf3d4d2e54e55c175c6b36a5f2375c3ebe835f6f95e3bR275

marshtompsxd commented on src/kubernetes_api_objects/spec/object_meta.rs:
```diff
@@ -80,6 +80,29 @@ impl ObjectMetaView {
         }
     }
 
+    pub open spec fn add_label(self, key: StringView, value: StringView) -> ObjectMetaView {
```
Why not call it `with_label`?

Catoverflow replied:
`with_labels` is for replacement, `add_label` and `remove_label` for incremental changes.

marshtompsxd commented on src/kubernetes_api_objects/exec/object_meta.rs:
```diff
@@ -214,6 +224,28 @@ impl ObjectMeta {
+    }
+
+    #[verifier(external_body)]
+    pub fn unset_label(&mut self, key: String)
```
How about calling it `remove_label`? Better to be consistent with `add_label`.

marshtompsxd commented on src/kubernetes_api_objects/exec/object_meta.rs:
```diff
@@ -186,6 +186,16 @@ impl ObjectMeta {
+        ensures b == self@.well_formed(),
+    {
+        self.inner.name.is_some()
+            && self.inner.namespace.is_some()
```
K8s has cluster-wide objects which do not have a `namespace`, tho we currently do not support (and do not need) these objects. I would suggest you move `self.inner.namespace.is_some()` to each CR object's `well_formed`, and then you can leave a comment here to explain the decision.

marshtompsxd commented on deploy/v2_vdeployment/v2_vdeployment.yaml:
```diff
@@ -0,0 +1,19 @@
+    spec:
+      containers:
+      - name: pause
+        image: k8s.gcr.io/pause:3.9
```
What does `pause` do? Just sleep?

Catoverflow replied:
Yes

marshtompsxd commented on src/kubernetes_api_objects/exec/label_selector.rs:
```diff
@@ -69,7 +70,9 @@ impl LabelSelector {
-        ensures res == self@.matches(labels@)
+        ensures
+            res == self@.matches(labels@),
+            res ==> self@.match_labels.is_None(),
```
I think the second postcondition is wrong. "Matches" does not mean the labels is None

marshtompsxd replied:
You should just remove this postcondition

marshtompsxd commented on src/v2/controllers/vdeployment_controller/model/reconciler.rs:
```diff
@@ -0,0 +1,278 @@
+        VDeploymentReconcileStepView::AfterGetReplicaSets => {
+            if !(resp_o.is_Some() && resp_o.get_Some_0().is_KResponse()
+            && resp_o.get_Some_0().get_KResponse_0().is_ListResponse()
+            && resp_o.get_Some_0().get_KResponse_0().get_ListResponse_0().res.is_ok()) {
```
There is some shortcut macro you can use here: https://github.com/anvil-verifier/anvil/blob/main/src/v2/reconciler/spec/io.rs#L70

The macro for list is not defined yet. You will need to define it.

Catoverflow replied:
I tried and the macros are not usable because of the borrowed value. The correct version is 
```
        if !(resp_o.is_some() && resp_o.as_ref().unwrap().is_k_response()
            && resp_o.as_ref().unwrap().as_k_response_ref().is_list_response()
            && resp_o.as_ref().unwrap().as_k_response_ref().as_list_response_ref().res.is_ok()) {
```
Which is not covered by macro

marshtompsxd commented on src/v2/controllers/vdeployment_controller/model/reconciler.rs:
```diff
@@ -0,0 +1,278 @@
+                    let (new_vrs_list, old_vrs_list) = filter_old_and_new_vrs(filter_vrs_list(vrs_list_or_none.unwrap(), vd), vd);
+                    (
+                        VDeploymentReconcileState {
+                            reconcile_step: VDeploymentReconcileStepView::RollReplicas,
```
I assume that after you implement the second way of rolling, you are going to have a branch here and go to different next steps?

Catoverflow replied:
Yes, as commented now

marshtompsxd commented on src/v2/controllers/vdeployment_controller/model/reconciler.rs:
```diff
@@ -0,0 +1,278 @@
+                if !state.new_vrs.unwrap().well_formed() {
+                    (error_state(state), None)
+                } else {
+                    let new_vrs = state.new_vrs.get_Some_0(); // k8s deterministically chooses the "oldest" vrs, we just use the first one
```
Can you put a `TODO` here for `k8s deterministically chooses the "oldest" vrs, we just use the first one`?

marshtompsxd commented on src/v2/controllers/vdeployment_controller/model/reconciler.rs:
```diff
@@ -0,0 +1,278 @@
+                    (error_state(state), None)
+                } else {
+                    let new_vrs = state.new_vrs.get_Some_0(); // k8s deterministically chooses the "oldest" vrs, we just use the first one
+                    let existing_replicas = new_vrs.spec.replicas.unwrap_or(1);
```
Can you rename `existing_replicas` to `current_replicas`?

marshtompsxd commented on src/v2/controllers/vdeployment_controller/model/reconciler.rs:
```diff
@@ -0,0 +1,278 @@
+                            obj: updated_new_vrs.marshal(),
+                        });
+                        let state_prime = VDeploymentReconcileState {
+                            reconcile_step: VDeploymentReconcileStepView::RollReplicas,
```
Why do we multiplex the `RollReplicas` step so much? Why not have different steps for managing new vrs and old vrs separately? The current way of representing steps makes it a bit hard to understand the state transition, and will create some difficulty in proving termination.

Catoverflow replied:
Fixed in the most recent changes. Now we have different states for different uses.

Catoverflow commented:
@marshtompsxd I believe we have another [flaky error](https://github.com/anvil-verifier/anvil/actions/runs/14986228462/job/42100589351?pr=611) here for VReplicaSet:

```
error: assertion failed
    --> v2/controllers/vreplicaset_controller/proof/helper_invariants/proof.rs:1224:120
     |
1224 | ...ey(key) && stronger_next(s, s_prime) implies requirements(key, s_prime) by {
     |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ assertion failed
```

It's inside the proof of `lemma_eventually_always_each_vrs_in_reconcile_implies_filtered_pods_owned_by_vrs`. However, if I run manually:
```
verus -L dependency=src/deps_hack/target/debug/deps --extern=deps_hack="src/deps_hack/target/debug/libdeps_hack.rlib" --verify-module vreplicaset_controller::proof::helper_invariants::proof --verify-function lemma_eventually_always_each_vrs_in_reconcile_implies_filtered_pods_owned_by_vrs src/v2_vreplicaset_controller.rs --multiple-errors 100 --expand-errors
```
It succeeded with no error. @codyjrivera Can you help have a look at this? I think we encountered similar flaky proofs before.

codyjrivera commented:
Try `#[verifier(spinoff_prover)]` first. If that doesn't work, ping me and then I'll need to put debugging effort into it.

Catoverflow commented:
@codyjrivera It works. Thanks for your help! We can figure out how to simplify it later.

> Try #[verifier(spinoff_prover)] first. If that doesn't work, ping me and then someone will need to put debugging effort into it.

Catoverflow commented:
Finally this monster is ready for review again

marshtompsxd commented on src/v2/controllers/vdeployment_controller/trusted/exec_types.rs:
@Catoverflow You can use https://github.com/anvil-verifier/anvil/blob/main/src/kubernetes_api_objects/exec/resource.rs#L86 to define VDeployment

Catoverflow replied:
I think these wrapper macros can reduce the duplication, but we still need to define other VDeployment-specific fields using the same structure. To me it's slightly better to use the same struct to define all methods as it improves readability

marshtompsxd commented on src/v2/controllers/vdeployment_controller/trusted/liveness_theorem.rs:
```diff
@@ -0,0 +1,28 @@
+        &&& old_vrs_list.len() == 0
+        &&& new_vrs_list[0].spec.replicas.unwrap_or(1) == vd.spec.replicas.unwrap_or(1)
+        &&& match_template_without_hash(vd, new_vrs_list[0])
+        &&& current_state_matches(new_vrs_list[0])
```
You can remove this last line

marshtompsxd commented on src/v2/controllers/vdeployment_controller/trusted/liveness_theorem.rs:
```diff
@@ -0,0 +1,28 @@
+
+// draft of ESR for VDeployment
+// TODO: add another version which talks about pods and derives from VRS ESR and this ESR
+pub open spec fn current_state_matches(vd: VDeploymentView) -> StatePred<ClusterState> {
```
Can you confirm that k8s deployment controller eventually deletes all old vrs? Or does it keep some old vrs but set their replicas to 0? If all old vrs are deleted eventually, you should just say vrs_list == Seq![one_new_vrs]; if not, you should say there is exactly one new vrs (whose template matches) and all other vrs' replicas are 0.

Catoverflow replied:
> Can you confirm that k8s deployment controller eventually deletes all old vrs?
It's configured in `d.Spec.RevisionHistoryLimit`, as the comment before `src/v2/controllers/vdeployment_controller/exec/reconciler.rs::reconcile_core` suggests

marshtompsxd commented on src/v2/controllers/vdeployment_controller/trusted/liveness_theorem.rs:
```diff
@@ -0,0 +1,28 @@
+use crate::temporal_logic::defs::*;
+use crate::vreplicaset_controller::trusted::spec_types::*;
+use crate::vdeployment_controller::trusted::spec_types::*;
+use crate::vdeployment_controller::model::reconciler::*;
```
Ideally, we should not import `crate::vdeployment_controller::model`. This file is the spec and should be read and understood by anyone using the controller, and `crate::vdeployment_controller::model` is part of the proof and only developers have to know about it.

Catoverflow replied:
But practically we will have 2 copies of utilities functions and need to prove the consistency between them like we once did for VRS, I'm weakly against this practice.

marshtompsxd replied:
You don't need to prove the consistency. You can just move them here, and let the model file import the liveness_theorem file. We don't have to do it in this PR.

marshtompsxd commented on src/kubernetes_api_objects/exec/object_meta.rs:
```diff
@@ -186,6 +186,15 @@ impl ObjectMeta {
     }
 
+    #[verifier(external_body)]
+    pub fn well_formed(&self) -> (b: bool)
```
Thinking twice, could you name it as `well_formed_for_namespaced` and then get `self.metadata().namespace().is_some()` back to this function? I'm sorry for the back-and-forth requests!

Catoverflow replied:
I think it's better to put it in another PR, and keep the current `vd.well_formed` as is.

marshtompsxd commented on src/v2/controllers/vreplicaset_controller/trusted/exec_types.rs:
You can use https://github.com/anvil-verifier/anvil/blob/main/src/kubernetes_api_objects/exec/resource.rs#L86 here

marshtompsxd commented on src/v2/controllers/vdeployment_controller/trusted/spec_types.rs:
```diff
@@ -127,9 +125,11 @@ impl ResourceView for VDeploymentView {
         &&& self.spec.template.is_Some()
         &&& self.spec.template.get_Some_0().metadata.is_Some()
+        // can be derived from selector match labels
+        &&& self.spec.template.get_Some_0().metadata.get_Some_0().labels.is_Some()
```
I think deployment's template should not be option: https://github.com/Arnavion/k8s-openapi/blob/master/src/v1_30/api/apps/v1/deployment_spec.rs#L28

marshtompsxd commented on src/v2/controllers/vreplicaset_controller/trusted/spec_types.rs:
```diff
@@ -152,4 +171,35 @@ pub struct VReplicaSetSpecView {
     pub template: Option<PodTemplateSpecView>,
```
This should not be Option (my bad for defining it this way). We should use this as reference: https://github.com/Arnavion/k8s-openapi/blob/master/src/v1_30/api/apps/v1/deployment_spec.rs#L28

marshtompsxd commented on src/v2/controllers/vdeployment_controller/exec/reconciler.rs:
```diff
@@ -0,0 +1,663 @@
+            }
+            let (new_vrs_list, old_vrs_list) = filter_old_and_new_vrs(filter_vrs_list(vrs_list_or_none.clone().unwrap(), vd), vd);
+            // no .last().cloned() in verus because "The verifier does not yet support the following Rust feature: overloaded deref"
+            let state = VDeploymentReconcileState {
```
Here you bind the variable `state` to a different value than the original argument

marshtompsxd commented on src/v2/controllers/vdeployment_controller/exec/reconciler.rs:
```diff
@@ -0,0 +1,663 @@
+                let (state_prime, req) = create_new_vrs(state, vd);
+                assert(state_prime@ == model_reconciler::create_new_vrs(state@, vd@).0);
+                assert(req.deep_view() == model_reconciler::create_new_vrs(state@, vd@).1);
+                assert(state_prime@ == model_reconciler::reconcile_core(vd@, resp_o.deep_view(), state@).0) by {
```
...and here you state that state_prime@ is the same as feeding the state@ to reconcile_core at model level. I don't think this assertion makes sense.

marshtompsxd commented on src/v2/controllers/vdeployment_controller/exec/reconciler.rs:
```diff
@@ -0,0 +1,644 @@
+}
+
+#[verifier(external_body)]
+pub fn reconcile_init_state() -> (state: VDeploymentReconcileState)
```
Why isn't this one verified?

Catoverflow replied:
Fixed, historical problem back that time.

marshtompsxd commented on error.log:
Can you remove the error log?

Catoverflow commented:
Let me fix e2e test

marshtompsxd commented:
@Catoverflow Somehow the CI gets stuck even before starts. Can you make whatever no-op changes and push again to trigger CI?

Catoverflow commented:
@marshtompsxd CI is run but not showed in this PR page. The commit number and branch name are correct.
see https://github.com/anvil-verifier/anvil/actions/runs/15050205275
